Resent50 Notebook

farazkh80 opened the pull request.

parthchadha commented:
1. Lets remove TP prefix from all modules.
2. There is no training with tripy so you can remove the `if training` code from BatchNorm.

parthchadha commented:
```
def get_tp_pred(model, image, processor):
    processed_image = processor(image, return_tensors="np")['pixel_values']
    tp_image = tp.Tensor(processed_image.squeeze(), dtype=tp.float32, device=tp.device("gpu"))
    tp_image = tp.unsqueeze(tp_image, dim=0)
    return tp.argmax(model(tp_image), -1).tolist()[0]
```

^ uses Tripy outside of the compiled model inference. Even though this is functionally correct, running Tripy operations in eager mode can be expensive and I would rather use torch for pre and post processing operations.

farazkh80 commented:
Thanks for the comments @parthchadha.

I removed `TP` prefix and extra training mode code from `BatchNorm`. I also did the pre/post processing steps with `torch` to avoid `tripy` eager mode.

pranavm-nvidia commented on tripy/examples/resent50/tripy-res50.ipynb:
```diff
@@ -0,0 +1,599 @@
+    "\n",
+    "        return x\n",
+    "\n",
+    "class Identity(tp.Module):\n",
+    "    def __init__(self):\n",
+    "        super(Identity, self).__init__()\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return x\n",
```
Do we need an identity module?

farazkh80 replied:
removed

pranavm-nvidia commented on tripy/examples/resent50/tripy-res50.ipynb:
```diff
@@ -0,0 +1,599 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MaxPool2d(tp.Module):\n",
+    "    def __init__(self, kernel_dims, stride=None, padding=((0, 0), (0, 0))):\n",
+    "        super(MaxPool2d, self).__init__()\n",
+    "        self.kernel_dims = kernel_dims\n",
+    "        self.stride = stride if stride is not None else kernel_dims\n",
+    "        self.padding = padding\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return tp.maxpool(x, kernel_dims=self.kernel_dims, stride=self.stride, padding=self.padding)\n",
```
Can we use `tp.maxpool` directly instead of wrapping it in a module?

farazkh80 replied:
done, removed module

pranavm-nvidia commented on tripy/examples/resent50/tripy-res50.ipynb:
```diff
@@ -0,0 +1,599 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class AdaptiveAvgPool2d(tp.Module):\n",
+    "    def __init__(self, output_size):\n",
+    "        super(AdaptiveAvgPool2d, self).__init__()\n",
+    "        self.output_size = output_size\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        N, C, H_in, W_in = x.shape\n",
+    "        H_out, W_out = self.output_size\n",
+    "\n",
+    "        # Calculate stride and kernel size\n",
+    "        stride_h = H_in // H_out\n",
+    "        stride_w = W_in // W_out\n",
+    "\n",
+    "        kernel_size_h = H_in - (H_out - 1) * stride_h\n",
+    "        kernel_size_w = W_in - (W_out - 1) * stride_w\n",
+    "\n",
+    "        return tp.avgpool(x, kernel_dims=(int(kernel_size_h), int(kernel_size_w)), stride=(int(stride_h), int(stride_w)))\n",
```
Could we just make this a helper function instead of a module?

farazkh80 replied:
done

pranavm-nvidia commented on tripy/examples/resent50/tripy-res50.ipynb:
```diff
@@ -0,0 +1,599 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Flatten(tp.Module):\n",
+    "    def __init__(self, start_dim=1, end_dim=-1):\n",
+    "        super(Flatten, self).__init__()\n",
+    "        self.start_dim = start_dim\n",
+    "        self.end_dim = end_dim\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return tp.flatten(x, start_dim=self.start_dim, end_dim=self.end_dim)\n",
```
Let's just use `tp.flatten` instead of creating a module?

farazkh80 replied:
done

farazkh80 commented:
Hey @pranavm-nvidia @parthchadha 

- Addressed the current suggested changes in the PR, and removed unnecessary modules from the notebook. 
- Added `BatchNorm` as a module similar to `LayerNorm` and `GroupNorm` under `module`

farazkh80 commented:
closing since I have a feature branch open now
